// dyna-showcase/src/Showcase/Showcase.tsx
import * as React from 'react';

import {IShowcase} from '../interfaces';

import {Menu} from './Menu/Menu';
import {Viewer} from './Viewer/Viewer';
import {IAppApi, IViewsAsLinks} from "../DynaShowcase";
import {cn} from "../utils/cn";

const styles = require('./Showcase.module.less');

export interface ShowcaseProps {
  showcase: IShowcase;
  viewSlug: string;
  propsSlug: string;
  appApi: IAppApi;
  menuStyle: any;
}

export interface ShowcaseState {
}

export class Showcase extends React.Component<ShowcaseProps, ShowcaseState> {
  public componentDidMount():void {
    const {viewSlug, appApi: {goTo, listOfLinkPaths}} = this.props;

    if (!viewSlug) {
      const firstLink: IViewsAsLinks = listOfLinkPaths[0];
      goTo(firstLink.viewSlug, firstLink.propsSlug);
    }
  }

  public render(): JSX.Element {
    const {
      showcase, viewSlug, propsSlug,
      appApi: {urlQuery: {hideMenu, zoom, showFrame, hideNavArrows, kbNav}},
    } = this.props;
    const asideClassName: string = cn(
      styles.asideMenu,
      hideMenu && styles.hideAsideMenu,
    );

    return (
      <div className={styles.container}>
        <div className={asideClassName}>
          <Menu
            showcase={showcase}
            viewSlug={viewSlug}
            propsSlug={propsSlug}
            appApi={this.props.appApi}
            style={this.props.menuStyle}
          />
        </div>
        <div className={styles.viewer}>
          <Viewer
            showcase={showcase}
            viewSlug={viewSlug}
            propsSlug={propsSlug}
            showFrame={!!showFrame}
            hideNavArrows={!!hideNavArrows}
            allowKeyboardNavigation={!!kbNav}
            zoom={Number(zoom)}
            appApi={this.props.appApi}
          />
        </div>
      </div>
    );
  }

}
